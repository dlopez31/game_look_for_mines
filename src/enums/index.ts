export enum CellEnum {
	Hidden = -1,
	Cero = 0,
	One = 1,
	Two = 2,
	Three = 3,
	Four = 4,
	Five = 5,
	Six = 6,
	Seven = 7,
	Eight = 8,
	Flag = 9,
	Mine = 10,
	ClickedMine = 11,
}

export enum ActionGame {
	SET_GRID = 'grid',
	SET_SCORE = 'score',
	SET_FACE = 'face',
	SET_START_GAME = 'startGame',
	SET_GAME_OVER = 'gameOver',
	SET_END_GAME = 'endGame',
}

export enum FacesEnum {
	DEFAULT = 'default',
	DOUBTFUL = 'doubtful',
	LOST = 'lost',
	WON = 'won',
}
