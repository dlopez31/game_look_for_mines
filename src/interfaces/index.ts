import { type CellEnum } from '../enums';

export type FaceType = 'default' | 'doubtful' | 'lost' | 'won';

export interface GameState {
	grid: CellEnum[][];
	face: FaceType;
	score: number;
	mines: number;
	booleanGrid: boolean[][];
	bombsOnGrid: number;
	currentMineRate: number;
	currentGridSize: number;
	lostStatus: boolean;
	isOpen: boolean;
	message: string;
	title: string;
}

export interface ICell {
	position: [number, number];
	cell: CellEnum;
	handleClickLeft: (position: [number, number]) => void;
	handleClickRight: (position: [number, number]) => void;
	handleOnMouseDown: () => void;
}

export interface Action {
	type: string;
	payload: any;
}

export interface AlertModalProps {
	title: string;
	message: string;
	onDismiss: () => void;
}

export interface UseGame {
	onFaceClick: () => void;
	handleOnclick: (position: [number, number]) => void;
	handleOnMouseDown: () => void;
	handleOnMouseUp: () => void;
	handleOnContextMenu: (position: [number, number]) => void;
	mines: number;
	score: number;
	face: string;
	grid: CellEnum[][];
	isOpen: boolean;
	message: string;
	title: string;
}
