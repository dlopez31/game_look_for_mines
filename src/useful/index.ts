import { type CellEnum } from '../enums';

export const bombQuantity = (
	percentage: number,
	currentGrid: CellEnum[][]
): number => Math.ceil(currentGrid.length * currentGrid.length * percentage);

export const booleanGridCreation = (size: number): boolean[][] => {
	const currentBooleanGrid = new Array(size);
	for (let i = 0; i < currentBooleanGrid.length; i++) {
		currentBooleanGrid[i] = new Array(size);
	}
	for (let i = 0; i < currentBooleanGrid.length; i++) {
		for (let j = 0; j < currentBooleanGrid.length; j++) {
			currentBooleanGrid[i][j] = false;
		}
	}
	return currentBooleanGrid;
};

export const bombPosition = (
	bombQ: number,
	currentBooleanGrid: boolean[][]
): boolean[][] => {
	let numerator = bombQ;
	let denominator = currentBooleanGrid.length * currentBooleanGrid.length;
	let bombActivation = false;
	for (let i = 0; i < currentBooleanGrid.length; i++) {
		for (let j = 0; j < currentBooleanGrid.length; j++) {
			bombActivation = !!(
				Math.random() <= numerator / denominator && numerator > 0
			);
			if (bombActivation) {
				numerator -= 1;
				denominator -= 1;
				currentBooleanGrid[i][j] = true;
			} else {
				denominator -= 1;
			}
			bombActivation = false;
		}
	}
	return currentBooleanGrid;
};

export const initialIterator = (index: number): number =>
	index > 0 ? index - 1 : 0;

export const lastIterator = (index: number, maxLength: number): number =>
	index === maxLength - 1 ? index : index + 1;

export const cellSelection = (
	rowPosition: number,
	columnPosition: number,
	currentBooleanGrid: boolean[][]
): number => {
	let cellImage: number = -10;
	if (currentBooleanGrid[rowPosition][columnPosition]) {
		cellImage = 11;
	} else {
		const rowInitiator = initialIterator(rowPosition);
		const rowStopper = lastIterator(rowPosition, currentBooleanGrid.length);
		const columnInitiator = initialIterator(columnPosition);
		const columnStopper = lastIterator(
			columnPosition,
			currentBooleanGrid.length
		);
		let bombCounter = 0;
		for (let i = rowInitiator; i <= rowStopper; i++) {
			for (let j = columnInitiator; j <= columnStopper; j++) {
				if (
					currentBooleanGrid[i][j] &&
					i < currentBooleanGrid.length &&
					j < currentBooleanGrid.length
				) {
					bombCounter += 1;
				}
			}
		}
		cellImage = bombCounter;
	}
	return cellImage;
};

export const getOpenCells = (grid: CellEnum[][]): number => {
	let openCells = 0;
	for (let i = 0; i < grid.length; i++) {
		for (let j = 0; j < grid[i].length; j++) {
			if (grid[i][j] >= 0 && grid[i][j] !== 9) {
				openCells = openCells + 1;
			}
		}
	}
	return openCells;
};

export const calculateScore = (
	currentBooleanGrid: boolean[][],
	counter: number,
	bombQ: number
): number =>
	Math.ceil(
		(100 / (currentBooleanGrid.length * currentBooleanGrid.length - bombQ)) *
			counter
	);

export const allOfUsAreDead = (
	rowPosition: number,
	columnPosition: number,
	currentGrid: CellEnum[][],
	currentBooleanGrid: boolean[][]
): CellEnum[][] => {
	for (let i = 0; i < currentGrid.length; i++) {
		for (let j = 0; j < currentGrid.length; j++) {
			if (i !== rowPosition || j !== columnPosition) {
				if (currentBooleanGrid[i][j]) {
					currentGrid[i][j] = 10;
				}
			}
		}
	}
	return currentGrid;
};

export const flagFeature = (
	bombQ: number,
	currentBooleanGrid: boolean[][],
	currentGrid: CellEnum[][]
): boolean => {
	let counter: number = 0;
	let totalFlags: number = 0;
	for (let i = 0; i < currentGrid.length; i++) {
		for (let j = 0; j < currentGrid.length; j++) {
			if (currentBooleanGrid[i][j] && currentGrid[i][j] === 9) {
				counter += 1;
			}
			if (currentGrid[i][j] === 9) {
				totalFlags += 1;
			}
		}
	}
	return !!(counter === bombQ && totalFlags === counter);
};
