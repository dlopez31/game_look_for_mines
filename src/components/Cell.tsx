import { CellEnum } from '../enums';
import { type ICell } from '../interfaces';

export const Cell = ({
	cell,
	position,
	handleClickLeft,
	handleOnMouseDown,
	handleClickRight,
}: ICell): JSX.Element => {
	const handleRightClick = (e: any, position: [number, number]): void => {
		e.preventDefault();
		handleClickRight(position);
	};

	const handleOnClick = (position: [number, number]): void => {
		handleClickLeft(position);
	};

	const renderCell = (): JSX.Element => {
		switch (cell) {
			case CellEnum.Hidden:
				return (
					<div
						className={'cell'}
						onMouseDown={handleOnMouseDown}
						onClick={(): void => {
							handleOnClick(position);
						}}
						onContextMenu={(e): void => {
							handleRightClick(e, position);
						}}
					></div>
				);
			case CellEnum.Flag:
				return (
					<div
						className={'cell cell__flag'}
						onMouseDown={handleOnMouseDown}
						onClick={(): void => {
							handleOnClick(position);
						}}
						onContextMenu={(e): void => {
							handleRightClick(e, position);
						}}
					></div>
				);
			case CellEnum.Mine:
				return <div className={'cell cell__mine cell__mine--shown'} />;
			case CellEnum.ClickedMine:
				return (
					<div
						className={'cell cell__mine cell__mine--shown cell__mine--clicked'}
					/>
				);
			default:
				return <div className={`cell cell--mines-${cell}`}>{cell}</div>;
		}
	};

	return renderCell();
};
