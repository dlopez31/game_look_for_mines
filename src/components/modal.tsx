import { Fragment } from 'react';
import { type AlertModalProps } from '../interfaces';

export const AlertModal = ({
	title,
	message,
	onDismiss,
}: AlertModalProps): JSX.Element => {
	return (
		<Fragment>
			<div className='fixed z-10 inset-0 bg-black opacity-50 dark:bg-gray-900'></div>
			<div className='md:animate-[bounce_8s_ease-in-out_infinite]  fixed z-20 inset-0 overflow-y-auto container mx-auto px-20'>
				<div className='flex items-center justify-center min-h-screen'>
					<div className='bg-[#202020] opacity-80 dark:bg-gray-800 rounded-2xl border-2 w-full'>
						<div className='px-6 py-4'>
							{/* Modal title */}
							<div className='text-8xl font-mono italic font-bold text-white'>
								{title}
							</div>
							<div className='mt-3 text-white text-5xl font-mono font-bold'>
								{message}
							</div>
						</div>
						<div className='w-xl flex justify-end px-6 py-4'>
							<button
								onClick={onDismiss}
								className='w-1/4 px-4 py-2 border-4 text-2xl font-bold bg-black text-white rounded-lg hover:bg-white hover:text-black focus:outline-none focus:ring-2 focus:ring-white'
							>
								Start Game
							</button>
						</div>
					</div>
				</div>
			</div>
		</Fragment>
	);
};
