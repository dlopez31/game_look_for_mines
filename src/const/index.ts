import { CellEnum, FacesEnum } from '../enums';
import { type GameState } from '../interfaces';

const size = 10;

export const initializeGrid = (): CellEnum[][] => {
	return Array.from({ length: size }, (_, i) => {
		return Array.from({ length: size }, (_, j) => {
			return CellEnum.Hidden;
		});
	});
};

export const initialize: GameState = {
	grid: initializeGrid(),
	face: FacesEnum.DEFAULT,
	score: 0,
	mines: 0,
	bombsOnGrid: 0,
	currentMineRate: 0.2,
	currentGridSize: size,
	isOpen: false,
	message: '',
	title: '',
};

export const MESSAGE_LOST = 'You lost try again';
export const MESSAGE_WON = 'Congratulations you have won';
export const TITLE_LOST = 'YOU LOST';
export const TITLE_WON = 'YOU WIN';
