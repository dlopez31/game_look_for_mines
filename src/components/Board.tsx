import { type CSSProperties } from 'react';
import { useGame } from '../hooks/useGame';
import { Cell } from './Cell';
import { AlertModal } from './modal';
import { type UseGame } from '../interfaces';
import '../styles.css';

export const Board = (): JSX.Element => {
	const {
		handleOnContextMenu,
		handleOnMouseDown,
		handleOnMouseUp,
		handleOnclick,
		onFaceClick,
		grid,
		mines,
		face,
		score,
		isOpen,
		message,
		title,
	}: UseGame = useGame();

	return (
		<div className='App' onMouseUp={handleOnMouseUp}>
			<section className='game-board'>
				<div className='controller'>
					<div className='mines-counter'>Mines: {mines}</div>
					<div
						onClick={onFaceClick}
						className={`status-button status-button--${face}-game`}
					></div>
					<div className='score'>Score: {score}</div>
				</div>
				{
					<div
						className='grid'
						style={{ '--grid-cells-side': grid.length } as CSSProperties}
					>
						{grid.length > 0 &&
							grid.flatMap((row, i) =>
								row.map((cell, j) => (
									<Cell
										key={`cell-${i}-${j}`}
										cell={cell}
										position={[i, j]}
										handleClickLeft={handleOnclick}
										handleClickRight={handleOnContextMenu}
										handleOnMouseDown={handleOnMouseDown}
									/>
								))
							)}
					</div>
				}
			</section>
			{isOpen && (
				<AlertModal
					title={title}
					message={message}
					onDismiss={onFaceClick}
				></AlertModal>
			)}
		</div>
	);
};
