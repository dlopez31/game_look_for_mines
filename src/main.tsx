import ReactDOM from 'react-dom/client';
import { Board } from './components/Board';

const divRoot = ReactDOM.createRoot(
	document.getElementById('root') as HTMLElement
);

divRoot.render(<Board />);
