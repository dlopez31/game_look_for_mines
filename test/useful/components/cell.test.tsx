import React from 'react';
import { render } from '@testing-library/react';
import { Cell } from '../../../src/components/Cell';
import { CellEnum } from '../../../src/enums';

describe('Cell', () => {
	it('renders correctly', () => {
		const { container } = render(
			<Cell
				cell={CellEnum.Hidden}
				position={[10, 10]}
				handleClickLeft={() => {}}
				handleOnMouseDown={() => {}}
				handleClickRight={() => {}}
			/>
		);
		expect(container.firstChild).toMatchSnapshot();
	});
});
