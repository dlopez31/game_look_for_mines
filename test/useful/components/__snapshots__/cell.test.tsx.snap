// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders correctly 1`] = `
<div
  class="cell"
/>
`;
